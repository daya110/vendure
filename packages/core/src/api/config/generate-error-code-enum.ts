import { buildSchema, extendSchema, GraphQLSchema, parse } from 'graphql';

export const ERROR_INTERFACE_NAME = 'ErrorResult';

/**
 * Generates the members of the `ErrorCode` enum dynamically, by getting the names of
 * all the types which inherit from the `ErrorResult` interface.
 */
export function generateErrorCodeEnum(typeDefsOrSchema: string | GraphQLSchema): GraphQLSchema {
    const schema = typeof typeDefsOrSchema === 'string' ? buildSchema(typeDefsOrSchema) : typeDefsOrSchema;
    const errorNodes = Object.values(schema.getTypeMap())
        .map(type => type.astNode)
        .filter(node => {
            return (
                node &&
                node?.kind === 'ObjectTypeDefinition' &&
                node.interfaces?.map(i => i.name.value).includes(ERROR_INTERFACE_NAME)
            );
        });
    if (!errorNodes.length) {
        return schema;
    }

    const errorCodeEnum = `
        extend enum ErrorCode {
            ${errorNodes.map(n => camelToUpperSnakeCase(n?.name.value || '')).join('\n')}
        }`;
    return extendSchema(schema, parse(errorCodeEnum));
}

function camelToUpperSnakeCase(input: string): string {
    return input.replace(/([a-z])([A-Z])/g, '$1_$2').toUpperCase();
}
