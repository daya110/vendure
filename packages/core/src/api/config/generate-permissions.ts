import { stitchSchemas, ValidationLevel } from '@graphql-tools/stitch';
import { GraphQLEnumType, GraphQLSchema } from 'graphql';
import { GraphQLEnumValueConfigMap } from 'graphql/type/definition';

import { getAllPermissionsMetadata } from '../../common/constants';
import { PermissionDefinition } from '../../common/permission-definition';

const PERMISSION_DESCRIPTION = `@description
Permissions for administrators and customers. Used to control access to
GraphQL resolvers via the {@link Allow} decorator.

## Understanding Permission.Owner

\`Permission.Owner\` is a special permission which is used in some of the Vendure resolvers to indicate that that resolver should only
be accessible to the "owner" of that resource.

For example, the Shop API \`activeCustomer\` query resolver should only return the Customer object for the "owner" of that Customer, i.e.
based on the activeUserId of the current session. As a result, the resolver code looks like this:

@example
\`\`\`TypeScript
\\@Query()
\\@Allow(Permission.Owner)
async activeCustomer(\\@Ctx() ctx: RequestContext): Promise<Customer | undefined> {
  const userId = ctx.activeUserId;
  if (userId) {
    return this.customerService.findOneByUserId(ctx, userId);
  }
}
\`\`\`

Here we can see that the "ownership" must be enforced by custom logic inside the resolver. Since "ownership" cannot be defined generally
nor statically encoded at build-time, any resolvers using \`Permission.Owner\` **must** include logic to enforce that only the owner
of the resource has access. If not, then it is the equivalent of using \`Permission.Public\`.


@docsCategory common`;

/**
 * Generates the `Permission` GraphQL enum based on the default & custom permission definitions.
 */
export function generatePermissionEnum(
    schema: GraphQLSchema,
    customPermissions: PermissionDefinition[],
): GraphQLSchema {
    const allPermissionsMetadata = getAllPermissionsMetadata(customPermissions);
    const values: GraphQLEnumValueConfigMap = {};
    let i = 0;
    for (const entry of allPermissionsMetadata) {
        values[entry.name] = {
            value: i,
            description: entry.description,
        };
        i++;
    }

    const permissionsEnum = new GraphQLEnumType({
        name: 'Permission',
        description: PERMISSION_DESCRIPTION,
        values,
    });

    return stitchSchemas({
        subschemas: [schema],
        types: [permissionsEnum],
        typeMergingOptions: { validationSettings: { validationLevel: ValidationLevel.Off } },
    });
}
