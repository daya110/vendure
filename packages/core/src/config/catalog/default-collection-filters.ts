import { LanguageCode } from '@vendure/common/lib/generated-types';

import { UserInputError } from '../../common/error/errors';
import { ProductVariant } from '../../entity/product-variant/product-variant.entity';

import { CollectionFilter } from './collection-filter';

/**
 * Filters for ProductVariants having the given facetValueIds (including parent Product)
 */
export const facetValueCollectionFilter = new CollectionFilter({
    args: {
        facetValueIds: {
            type: 'ID',
            list: true,
            ui: {
                component: 'facet-value-form-input',
            },
        },
        containsAny: { type: 'boolean' },
    },
    code: 'facet-value-filter',
    description: [{ languageCode: LanguageCode.en, value: 'Filter by FacetValues' }],
    apply: (qb, args) => {
        const ids = args.facetValueIds;

        if (ids.length) {
            const idsName = `ids_${ids.join('_')}`;
            const countName = `count_${ids.join('_')}`;
            const productFacetValues = qb.connection
                .createQueryBuilder(ProductVariant, 'product_variant')
                .select('product_variant.id', 'variant_id')
                .addSelect('facet_value.id', 'facet_value_id')
                .leftJoin('product_variant.facetValues', 'facet_value')
                .where(`facet_value.id IN (:...${idsName})`);

            const variantFacetValues = qb.connection
                .createQueryBuilder(ProductVariant, 'product_variant')
                .select('product_variant.id', 'variant_id')
                .addSelect('facet_value.id', 'facet_value_id')
                .leftJoin('product_variant.product', 'product')
                .leftJoin('product.facetValues', 'facet_value')
                .where(`facet_value.id IN (:...${idsName})`);

            const union = qb.connection
                .createQueryBuilder()
                .select('union_table.variant_id')
                .from(
                    `(${productFacetValues.getQuery()} UNION ${variantFacetValues.getQuery()})`,
                    'union_table',
                )
                .groupBy('variant_id')
                .having(`COUNT(*) >= :${countName}`);

            const variantIds = qb.connection
                .createQueryBuilder()
                .select('variant_ids_table.variant_id')
                .from(`(${union.getQuery()})`, 'variant_ids_table');

            qb.andWhere(`productVariant.id IN (${variantIds.getQuery()})`).setParameters({
                [idsName]: ids,
                [countName]: args.containsAny ? 1 : ids.length,
            });
        } else {
            // If no facetValueIds are specified, no ProductVariants will be matched.
            qb.andWhere('1 = 0');
        }
        return qb;
    },
});

export const variantNameCollectionFilter = new CollectionFilter({
    args: {
        operator: {
            type: 'string',
            ui: {
                component: 'select-form-input',
                options: [
                    { value: 'startsWith' },
                    { value: 'endsWith' },
                    { value: 'contains' },
                    { value: 'doesNotContain' },
                ],
            },
        },
        term: { type: 'string' },
    },
    code: 'variant-name-filter',
    description: [{ languageCode: LanguageCode.en, value: 'Filter by ProductVariant name' }],
    apply: (qb, args) => {
        qb.leftJoin('productVariant.translations', 'translation');
        const LIKE = qb.connection.options.type === 'postgres' ? 'ILIKE' : 'LIKE';
        switch (args.operator) {
            case 'contains':
                return qb.andWhere(`translation.name ${LIKE} :term`, { term: `%${args.term}%` });
            case 'doesNotContain':
                return qb.andWhere(`translation.name NOT ${LIKE} :term`, { term: `%${args.term}%` });
            case 'startsWith':
                return qb.andWhere(`translation.name ${LIKE} :term`, { term: `${args.term}%` });
            case 'endsWith':
                return qb.andWhere(`translation.name ${LIKE} :term`, { term: `%${args.term}` });
            default:
                throw new UserInputError(`${args.operator} is not a valid operator`);
        }
    },
});

export const defaultCollectionFilters = [facetValueCollectionFilter, variantNameCollectionFilter];
