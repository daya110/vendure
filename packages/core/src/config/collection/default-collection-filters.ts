import { ConfigArgType } from '@vendure/common/lib/generated-types';
import { Brackets } from 'typeorm';

import { CollectionFilter } from './collection-filter';

/**
 * Filters for ProductVariants having the given facetValueIds (including parent Product)
 */
export const facetValueCollectionFilter = new CollectionFilter({
    args: {
        facetValueIds: ConfigArgType.FACET_VALUE_IDS,
        containsAny: ConfigArgType.BOOLEAN,
    },
    code: 'facet-value-filter',
    description: 'Filter by FacetValues',
    apply: (qb, args) => {
        if (args.facetValueIds.length) {
            qb.leftJoin('productVariant.product', 'product')
                .leftJoin('product.facetValues', 'productFacetValues')
                .leftJoin('productVariant.facetValues', 'variantFacetValues')
                .andWhere(
                    new Brackets(qb1 => {
                        const ids = args.facetValueIds;
                        return qb1
                            .where(`productFacetValues.id IN (:...ids)`, {ids})
                            .orWhere(`variantFacetValues.id IN (:...ids)`, {ids});
                    }),
                )
                .groupBy('productVariant.id')
                .having(`COUNT(1) >= :count`, { count: args.containsAny ? 1 : args.facetValueIds.length });
        } else {
            // If no facetValueIds are specified, no ProductVariants will be matched.
            qb.andWhere('1 = 0');
        }
        return qb;
    },
});

export const variantNameCollectionFilter = new CollectionFilter({
    args: {
        operator: ConfigArgType.STRING_OPERATOR,
        term: ConfigArgType.STRING,
    },
    code: 'variant-name-filter',
    description: 'Filter by ProductVariant name',
    apply: (qb, args) => {
        qb.leftJoin('productVariant.translations', 'translation');
        switch (args.operator) {
            case 'contains':
                return qb.andWhere('translation.name LIKE :term', { term: `%${args.term}%` });
            case 'doesNotContain':
                return qb.andWhere('translation.name NOT LIKE :term', { term: `%${args.term}%` });
            case 'startsWith':
                return qb.andWhere('translation.name LIKE :term', { term: `${args.term}%` });
            case 'endsWith':
                return qb.andWhere('translation.name LIKE :term', { term: `%${args.term}` });
        }
    },
});
