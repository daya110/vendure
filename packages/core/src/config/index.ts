export * from './asset-naming-strategy/asset-naming-strategy';
export * from './asset-naming-strategy/default-asset-naming-strategy';
export * from './asset-preview-strategy/asset-preview-strategy';
export * from './asset-storage-strategy/asset-storage-strategy';
export * from './asset-storage-strategy/local-asset-storage-strategy';
export * from './entity-id-strategy/auto-increment-id-strategy';
export * from './config.service';
export * from './entity-id-strategy/entity-id-strategy';
export * from './entity-id-strategy/uuid-id-strategy';
export * from './order-merge-strategy/order-merge-strategy';
export * from './logger/vendure-logger';
export * from './logger/default-logger';
export * from './logger/noop-logger';
export * from './payment-method/example-payment-method-config';
export * from './payment-method/payment-method-handler';
export * from './promotion/default-promotion-actions';
export * from './promotion/default-promotion-conditions';
export * from './promotion/promotion-action';
export * from './promotion/promotion-condition';
export * from './shipping-method/default-shipping-calculator';
export * from './shipping-method/default-shipping-eligibility-checker';
export * from './shipping-method/shipping-calculator';
export * from './shipping-method/shipping-eligibility-checker';
export * from './vendure-config';
export * from './merge-config';
