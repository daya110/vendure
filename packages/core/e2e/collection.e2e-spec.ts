/* tslint:disable:no-non-null-assertion */
import {
    Collection,
    ConfigArgType,
    CreateCollection,
    CreateCollectionInput,
    FacetValue,
    GetAssetList,
    GetCollection,
    LanguageCode,
    MoveCollection,
    ProductWithVariants,
    SortOrder,
    UpdateCollection,
    UpdateProduct,
    UpdateProductVariants,
} from '@vendure/common/lib/generated-types';
import { ROOT_COLLECTION_NAME } from '@vendure/common/lib/shared-constants';
import gql from 'graphql-tag';
import path from 'path';

import {
    CREATE_COLLECTION,
    GET_COLLECTION,
    MOVE_COLLECTION,
    UPDATE_COLLECTION,
} from '../../../admin-ui/src/app/data/definitions/collection-definitions';
import { FACET_VALUE_FRAGMENT } from '../../../admin-ui/src/app/data/definitions/facet-definitions';
import { GET_ASSET_LIST, UPDATE_PRODUCT, UPDATE_PRODUCT_VARIANTS } from '../../../admin-ui/src/app/data/definitions/product-definitions';
import { facetValueCollectionFilter } from '../src/config/collection/default-collection-filters';

import { TEST_SETUP_TIMEOUT_MS } from './config/test-config';
import { TestAdminClient } from './test-client';
import { TestServer } from './test-server';
import { assertThrowsWithMessage } from './utils/assert-throws-with-message';

describe('Collection resolver', () => {
    const client = new TestAdminClient();
    const server = new TestServer();
    let assets: GetAssetList.Items[];
    let facetValues: FacetValue.Fragment[];
    let electronicsCollection: Collection.Fragment;
    let computersCollection: Collection.Fragment;
    let pearCollection: Collection.Fragment;

    beforeAll(async () => {
        const token = await server.init({
            productsCsvPath: path.join(__dirname, 'fixtures/e2e-products-collections.csv'),
            customerCount: 1,
        });
        await client.init();
        const assetsResult = await client.query<GetAssetList.Query, GetAssetList.Variables>(GET_ASSET_LIST, {
            options: {
                sort: {
                    name: SortOrder.ASC,
                },
            },
        });
        assets = assetsResult.assets.items;
        const facetValuesResult = await client.query(GET_FACET_VALUES);
        facetValues = facetValuesResult.facets.items.reduce(
            (values: any, facet: any) => [...values, ...facet.values],
            [],
        );
    }, TEST_SETUP_TIMEOUT_MS);

    afterAll(async () => {
        await server.destroy();
    });

    describe('createCollection', () => {
        it('creates a root collection', async () => {
            const result = await client.query<CreateCollection.Mutation, CreateCollection.Variables>(
                CREATE_COLLECTION,
                {
                    input: {
                        assetIds: [assets[0].id, assets[1].id],
                        featuredAssetId: assets[1].id,
                        filters: [
                            {
                                code: facetValueCollectionFilter.code,
                                arguments: [
                                    {
                                        name: 'facetValueIds',
                                        value: `["${getFacetValueId('electronics')}"]`,
                                        type: ConfigArgType.FACET_VALUE_IDS,
                                    },
                                ],
                            },
                        ],
                        translations: [
                            { languageCode: LanguageCode.en, name: 'Electronics', description: '' },
                        ],
                    },
                },
            );

            electronicsCollection = result.createCollection;
            expect(electronicsCollection).toMatchSnapshot();
            expect(electronicsCollection.parent!.name).toBe(ROOT_COLLECTION_NAME);
        });

        it('creates a nested category', async () => {
            const result = await client.query<CreateCollection.Mutation, CreateCollection.Variables>(
                CREATE_COLLECTION,
                {
                    input: {
                        parentId: electronicsCollection.id,
                        translations: [{ languageCode: LanguageCode.en, name: 'Computers', description: '' }],
                        filters: [
                            {
                                code: facetValueCollectionFilter.code,
                                arguments: [
                                    {
                                        name: 'facetValueIds',
                                        value: `["${getFacetValueId('computers')}"]`,
                                        type: ConfigArgType.FACET_VALUE_IDS,
                                    },
                                ],
                            },
                        ],
                    },
                },
            );
            computersCollection = result.createCollection;
            expect(computersCollection.parent!.name).toBe(electronicsCollection.name);
        });

        it('creates a 2nd level nested category', async () => {
            const result = await client.query<CreateCollection.Mutation, CreateCollection.Variables>(
                CREATE_COLLECTION,
                {
                    input: {
                        parentId: computersCollection.id,
                        translations: [{ languageCode: LanguageCode.en, name: 'Pear', description: '' }],
                        filters: [
                            {
                                code: facetValueCollectionFilter.code,
                                arguments: [
                                    {
                                        name: 'facetValueIds',
                                        value: `["${getFacetValueId('pear')}"]`,
                                        type: ConfigArgType.FACET_VALUE_IDS,
                                    },
                                ],
                            },
                        ],
                    },
                },
            );
            pearCollection = result.createCollection;
            expect(pearCollection.parent!.name).toBe(computersCollection.name);
        });
    });

    it('collection query', async () => {
        const result = await client.query<GetCollection.Query, GetCollection.Variables>(GET_COLLECTION, {
            id: computersCollection.id,
        });
        if (!result.collection) {
            fail(`did not return the collection`);
            return;
        }
        expect(result.collection.id).toBe(computersCollection.id);
    });

    it('breadcrumbs', async () => {
        const result = await client.query(GET_COLLECTION_BREADCRUMBS, {
            id: pearCollection.id,
        });
        if (!result.collection) {
            fail(`did not return the collection`);
            return;
        }
        expect(result.collection.breadcrumbs).toEqual([
            { id: 'T_1', name: ROOT_COLLECTION_NAME },
            { id: electronicsCollection.id, name: electronicsCollection.name },
            { id: computersCollection.id, name: computersCollection.name },
            { id: pearCollection.id, name: pearCollection.name },
        ]);
    });

    it('breadcrumbs for root collection', async () => {
        const result = await client.query(GET_COLLECTION_BREADCRUMBS, {
            id: 'T_1',
        });
        if (!result.collection) {
            fail(`did not return the collection`);
            return;
        }
        expect(result.collection.breadcrumbs).toEqual([{ id: 'T_1', name: ROOT_COLLECTION_NAME }]);
    });

    it('updateCollection', async () => {
        const result = await client.query<UpdateCollection.Mutation, UpdateCollection.Variables>(
            UPDATE_COLLECTION,
            {
                input: {
                    id: pearCollection.id,
                    assetIds: [assets[1].id],
                    featuredAssetId: assets[1].id,
                    translations: [{ languageCode: LanguageCode.en, description: 'Apple stuff ' }],
                },
            },
        );

        expect(result.updateCollection).toMatchSnapshot();
    });

    describe('moveCollection', () => {
        it('moves a collection to a new parent', async () => {
            const result = await client.query<MoveCollection.Mutation, MoveCollection.Variables>(
                MOVE_COLLECTION,
                {
                    input: {
                        collectionId: pearCollection.id,
                        parentId: electronicsCollection.id,
                        index: 0,
                    },
                },
            );

            expect(result.moveCollection.parent!.id).toBe(electronicsCollection.id);

            const positions = await getChildrenOf(electronicsCollection.id);
            expect(positions.map((i: any) => i.id)).toEqual([pearCollection.id, computersCollection.id]);
        });

        it('re-evaluates Collection contents on move', async () => {
            const result = await client.query(GET_COLLECTION_PRODUCT_VARIANTS, { id: pearCollection.id });
            expect(result.collection.productVariants.items.map((i: any) => i.name)).toEqual([
                'Laptop 13 inch 8GB',
                'Laptop 15 inch 8GB',
                'Laptop 13 inch 16GB',
                'Laptop 15 inch 16GB',
                'Instant Camera',
            ]);
        });

        it('alters the position in the current parent', async () => {
            await client.query<MoveCollection.Mutation, MoveCollection.Variables>(MOVE_COLLECTION, {
                input: {
                    collectionId: pearCollection.id,
                    parentId: electronicsCollection.id,
                    index: 1,
                },
            });

            const afterResult = await getChildrenOf(electronicsCollection.id);
            expect(afterResult.map((i: any) => i.id)).toEqual([computersCollection.id, pearCollection.id]);
        });

        it('corrects an out-of-bounds negative index value', async () => {
            await client.query<MoveCollection.Mutation, MoveCollection.Variables>(MOVE_COLLECTION, {
                input: {
                    collectionId: pearCollection.id,
                    parentId: electronicsCollection.id,
                    index: -3,
                },
            });

            const afterResult = await getChildrenOf(electronicsCollection.id);
            expect(afterResult.map((i: any) => i.id)).toEqual([pearCollection.id, computersCollection.id]);
        });

        it('corrects an out-of-bounds positive index value', async () => {
            await client.query<MoveCollection.Mutation, MoveCollection.Variables>(MOVE_COLLECTION, {
                input: {
                    collectionId: pearCollection.id,
                    parentId: electronicsCollection.id,
                    index: 10,
                },
            });

            const afterResult = await getChildrenOf(electronicsCollection.id);
            expect(afterResult.map((i: any) => i.id)).toEqual([computersCollection.id, pearCollection.id]);
        });

        it(
            'throws if attempting to move into self',
            assertThrowsWithMessage(
                () =>
                    client.query<MoveCollection.Mutation, MoveCollection.Variables>(MOVE_COLLECTION, {
                        input: {
                            collectionId: pearCollection.id,
                            parentId: pearCollection.id,
                            index: 0,
                        },
                    }),
                `Cannot move a Collection into itself`,
            ),
        );

        it(
            'throws if attempting to move into a decendant of self',
            assertThrowsWithMessage(
                () =>
                    client.query<MoveCollection.Mutation, MoveCollection.Variables>(MOVE_COLLECTION, {
                        input: {
                            collectionId: pearCollection.id,
                            parentId: pearCollection.id,
                            index: 0,
                        },
                    }),
                `Cannot move a Collection into itself`,
            ),
        );

        async function getChildrenOf(parentId: string): Promise<Array<{ name: string; id: string }>> {
            const result = await client.query(GET_COLLECTIONS);
            return result.collections.items.filter((i: any) => i.parent.id === parentId);
        }
    });

    describe('filters', () => {
        it('Collection with no filters has no productVariants', async () => {
            const result = await client.query(CREATE_COLLECTION_SELECT_VARIANTS, {
                input: {
                    translations: [{ languageCode: LanguageCode.en, name: 'Empty', description: '' }],
                    filters: [],
                } as CreateCollectionInput,
            });
            expect(result.createCollection.productVariants.totalItems).toBe(0);
        });

        describe('facetValue filter', () => {
            it('electronics', async () => {
                const result = await client.query(GET_COLLECTION_PRODUCT_VARIANTS, {
                    id: electronicsCollection.id,
                });
                expect(result.collection.productVariants.items.map((i: any) => i.name)).toEqual([
                    'Laptop 13 inch 8GB',
                    'Laptop 15 inch 8GB',
                    'Laptop 13 inch 16GB',
                    'Laptop 15 inch 16GB',
                    'Curvy Monitor 24 inch',
                    'Curvy Monitor 27 inch',
                    'Gaming PC i7-8700 240GB SSD',
                    'Gaming PC R7-2700 240GB SSD',
                    'Gaming PC i7-8700 120GB SSD',
                    'Gaming PC R7-2700 120GB SSD',
                    'Hard Drive 1TB',
                    'Hard Drive 2TB',
                    'Hard Drive 3TB',
                    'Hard Drive 4TB',
                    'Hard Drive 6TB',
                    'Clacky Keyboard',
                    'USB Cable',
                    'Instant Camera',
                    'Camera Lens',
                    'Tripod',
                    'SLR Camera',
                ]);
            });

            it('computers', async () => {
                const result = await client.query(GET_COLLECTION_PRODUCT_VARIANTS, {
                    id: computersCollection.id,
                });
                expect(result.collection.productVariants.items.map((i: any) => i.name)).toEqual([
                    'Laptop 13 inch 8GB',
                    'Laptop 15 inch 8GB',
                    'Laptop 13 inch 16GB',
                    'Laptop 15 inch 16GB',
                    'Curvy Monitor 24 inch',
                    'Curvy Monitor 27 inch',
                    'Gaming PC i7-8700 240GB SSD',
                    'Gaming PC R7-2700 240GB SSD',
                    'Gaming PC i7-8700 120GB SSD',
                    'Gaming PC R7-2700 120GB SSD',
                    'Hard Drive 1TB',
                    'Hard Drive 2TB',
                    'Hard Drive 3TB',
                    'Hard Drive 4TB',
                    'Hard Drive 6TB',
                    'Clacky Keyboard',
                    'USB Cable',
                ]);
            });

            it('photo and pear', async () => {
                const result = await client.query(CREATE_COLLECTION_SELECT_VARIANTS, {
                    input: {
                        translations: [
                            { languageCode: LanguageCode.en, name: 'Photo Pear', description: '' },
                        ],
                        filters: [
                            {
                                code: facetValueCollectionFilter.code,
                                arguments: [
                                    {
                                        name: 'facetValueIds',
                                        value: `["${getFacetValueId('pear')}", "${getFacetValueId(
                                            'photo',
                                        )}"]`,
                                        type: ConfigArgType.FACET_VALUE_IDS,
                                    },
                                ],
                            },
                        ],
                    } as CreateCollectionInput,
                });
                expect(result.createCollection.productVariants.items.map((i: any) => i.name)).toEqual([
                    'Instant Camera',
                ]);
            });
        });

        describe('re-evaluation of contents on changes', () => {
            let products: ProductWithVariants.Fragment[];

            beforeAll(async () => {
                const result = await client.query(gql`
                    query {
                        products {
                            items {
                                id
                                name
                                variants {
                                    id
                                }
                            }
                        }
                    }
                `);
                products = result.products.items;
            });

            it('updates contents when Product is updated', async () => {
                await client.query<UpdateProduct.Mutation, UpdateProduct.Variables>(UPDATE_PRODUCT, {
                    input: {
                        id: products[1].id,
                        facetValueIds: [
                            getFacetValueId('electronics'),
                            getFacetValueId('computers'),
                            getFacetValueId('pear'),
                        ],
                    },
                });

                const result = await client.query(GET_COLLECTION_PRODUCT_VARIANTS, { id: pearCollection.id });
                expect(result.collection.productVariants.items.map((i: any) => i.name)).toEqual([
                    'Laptop 13 inch 8GB',
                    'Laptop 15 inch 8GB',
                    'Laptop 13 inch 16GB',
                    'Laptop 15 inch 16GB',
                    'Curvy Monitor 24 inch',
                    'Curvy Monitor 27 inch',
                    'Instant Camera',
                ]);
            });

            it('updates contents when ProductVariant is updated', async () => {
                const gamingPcFirstVariant = products.find(p => p.name === 'Gaming PC')!.variants[0];
                await client.query<UpdateProductVariants.Mutation, UpdateProductVariants.Variables>(
                    UPDATE_PRODUCT_VARIANTS,
                    {
                        input: [
                            {
                                id: gamingPcFirstVariant.id,
                                facetValueIds: [getFacetValueId('pear')],
                            },
                        ],
                    },
                );
                const result = await client.query(GET_COLLECTION_PRODUCT_VARIANTS, { id: pearCollection.id });
                expect(result.collection.productVariants.items.map((i: any) => i.name)).toEqual([
                    'Laptop 13 inch 8GB',
                    'Laptop 15 inch 8GB',
                    'Laptop 13 inch 16GB',
                    'Laptop 15 inch 16GB',
                    'Curvy Monitor 24 inch',
                    'Curvy Monitor 27 inch',
                    'Gaming PC i7-8700 240GB SSD',
                    'Instant Camera',
                ]);
            });

            it('correctly filters when ProductVariant and Product both have matching FacetValue', async () => {
                const gamingPcFirstVariant = products.find(p => p.name === 'Gaming PC')!.variants[0];
                await client.query<UpdateProductVariants.Mutation, UpdateProductVariants.Variables>(
                    UPDATE_PRODUCT_VARIANTS,
                    {
                        input: [
                            {
                                id: gamingPcFirstVariant.id,
                                facetValueIds: [getFacetValueId('electronics'), getFacetValueId('pear')],
                            },
                        ],
                    },
                );
                const result = await client.query(GET_COLLECTION_PRODUCT_VARIANTS, { id: pearCollection.id });
                expect(result.collection.productVariants.items.map((i: any) => i.name)).toEqual([
                    'Laptop 13 inch 8GB',
                    'Laptop 15 inch 8GB',
                    'Laptop 13 inch 16GB',
                    'Laptop 15 inch 16GB',
                    'Curvy Monitor 24 inch',
                    'Curvy Monitor 27 inch',
                    'Gaming PC i7-8700 240GB SSD',
                    'Instant Camera',
                ]);
            });
        });
    });

    describe('Product collections property', () => {
        it('returns all collections to which the Product belongs', async () => {
            const result = await client.query(GET_COLLECTIONS_FOR_PRODUCTS, { term: 'camera' });
            expect(result.products.items[0].collections).toEqual([
                { id: 'T_3', name: 'Electronics' },
                { id: 'T_5', name: 'Pear' },
                { id: 'T_7', name: 'Photo Pear' },
            ]);
        });
    });

    function getFacetValueId(code: string): string {
        const match = facetValues.find(fv => fv.code === code);
        if (!match) {
            throw new Error(`Could not find a FacetValue with the code "${code}"`);
        }
        return match.id;
    }
});

const GET_FACET_VALUES = gql`
    query {
        facets {
            items {
                values {
                    ...FacetValue
                }
            }
        }
    }
    ${FACET_VALUE_FRAGMENT}
`;

const GET_COLLECTIONS = gql`
    query GetCollections {
        collections(languageCode: en) {
            items {
                id
                name
                position
                parent {
                    id
                    name
                }
            }
        }
    }
`;

const GET_COLLECTION_PRODUCT_VARIANTS = gql`
    query GetCollectionProducts($id: ID!) {
        collection(id: $id) {
            productVariants {
                items {
                    id
                    name
                    facetValues {
                        code
                    }
                }
            }
        }
    }
`;

const CREATE_COLLECTION_SELECT_VARIANTS = gql`
    mutation($input: CreateCollectionInput!) {
        createCollection(input: $input) {
            productVariants {
                items {
                    name
                }
                totalItems
            }
        }
    }
`;

const GET_COLLECTION_BREADCRUMBS = gql`
    query GetCollectionBreadcrumbs($id: ID!) {
        collection(id: $id) {
            breadcrumbs {
                id
                name
            }
        }
    }
`;

const GET_COLLECTIONS_FOR_PRODUCTS = gql`
    query GetCollectionsForProducts($term: String!) {
        products(options: { filter: { name: { contains: $term } } }) {
            items {
                id
                name
                collections {
                    id
                    name
                }
            }
        }
    }
`;
