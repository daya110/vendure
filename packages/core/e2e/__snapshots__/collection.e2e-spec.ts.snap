// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Collection resolver createCollection creates a root collection 1`] = `
Object {
  "assets": Array [
    Object {
      "fileSize": 1680,
      "id": "T_2",
      "mimeType": "image/jpeg",
      "name": "alexandru-acea-686569-unsplash.jpg",
      "preview": "test-url/test-assets/alexandru-acea-686569-unsplash__preview.jpg",
      "source": "test-url/test-assets/alexandru-acea-686569-unsplash.jpg",
      "type": "IMAGE",
    },
    Object {
      "fileSize": 1680,
      "id": "T_1",
      "mimeType": "image/jpeg",
      "name": "derick-david-409858-unsplash.jpg",
      "preview": "test-url/test-assets/derick-david-409858-unsplash__preview.jpg",
      "source": "test-url/test-assets/derick-david-409858-unsplash.jpg",
      "type": "IMAGE",
    },
  ],
  "children": Array [],
  "description": "",
  "featuredAsset": Object {
    "fileSize": 1680,
    "id": "T_1",
    "mimeType": "image/jpeg",
    "name": "derick-david-409858-unsplash.jpg",
    "preview": "test-url/test-assets/derick-david-409858-unsplash__preview.jpg",
    "source": "test-url/test-assets/derick-david-409858-unsplash.jpg",
    "type": "IMAGE",
  },
  "filters": Array [
    Object {
      "args": Array [
        Object {
          "name": "facetValueIds",
          "value": "[\\"T_1\\"]",
        },
        Object {
          "name": "containsAny",
          "value": "false",
        },
      ],
      "code": "facet-value-filter",
    },
  ],
  "id": "T_3",
  "isPrivate": false,
  "languageCode": "en",
  "name": "Electronics",
  "parent": Object {
    "id": "T_1",
    "name": "__root_collection__",
  },
  "slug": "electronics",
  "translations": Array [
    Object {
      "description": "",
      "id": "T_3",
      "languageCode": "en",
      "name": "Electronics",
      "slug": "electronics",
    },
  ],
}
`;

exports[`Collection resolver updateCollection updates with assets 1`] = `
Object {
  "assets": Array [
    Object {
      "fileSize": 1680,
      "id": "T_1",
      "mimeType": "image/jpeg",
      "name": "derick-david-409858-unsplash.jpg",
      "preview": "test-url/test-assets/derick-david-409858-unsplash__preview.jpg",
      "source": "test-url/test-assets/derick-david-409858-unsplash.jpg",
      "type": "IMAGE",
    },
    Object {
      "fileSize": 1680,
      "id": "T_3",
      "mimeType": "image/jpeg",
      "name": "florian-olivo-1166419-unsplash.jpg",
      "preview": "test-url/test-assets/florian-olivo-1166419-unsplash__preview.jpg",
      "source": "test-url/test-assets/florian-olivo-1166419-unsplash.jpg",
      "type": "IMAGE",
    },
  ],
  "children": Array [],
  "description": "Apple stuff ",
  "featuredAsset": Object {
    "fileSize": 1680,
    "id": "T_1",
    "mimeType": "image/jpeg",
    "name": "derick-david-409858-unsplash.jpg",
    "preview": "test-url/test-assets/derick-david-409858-unsplash__preview.jpg",
    "source": "test-url/test-assets/derick-david-409858-unsplash.jpg",
    "type": "IMAGE",
  },
  "filters": Array [
    Object {
      "args": Array [
        Object {
          "name": "facetValueIds",
          "value": "[\\"T_3\\"]",
        },
        Object {
          "name": "containsAny",
          "value": "false",
        },
      ],
      "code": "facet-value-filter",
    },
  ],
  "id": "T_5",
  "isPrivate": false,
  "languageCode": "en",
  "name": "Pear",
  "parent": Object {
    "id": "T_4",
    "name": "Computers",
  },
  "slug": "apple-stuff",
  "translations": Array [
    Object {
      "description": "Apple stuff ",
      "id": "T_5",
      "languageCode": "en",
      "name": "Pear",
      "slug": "apple-stuff",
    },
  ],
}
`;
